internal/edgeapi/backend/k8s/backend.go

```diff
@@ -0,0 +1,217 @@
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/project-flotta/flotta-operator/api/v1alpha1"
+	"github.com/project-flotta/flotta-operator/internal/common/utils"
+	backendapi "github.com/project-flotta/flotta-operator/internal/edgeapi/backend"
+	"github.com/project-flotta/flotta-operator/internal/edgeapi/hardware"
+	"github.com/project-flotta/flotta-operator/models"
+	"github.com/project-flotta/flotta-operator/pkg/mtls"
+)
+
+const (
+	YggdrasilConnectionFinalizer = "yggdrasil-connection-finalizer"
+	YggdrasilWorkloadFinalizer   = "yggdrasil-workload-finalizer"
+
+	AuthzKey mtls.RequestAuthKey = "APIAuthzkey"
+)
+
+type backend struct {
+	logger           *zap.SugaredLogger
+	repository       RepositoryFacade
+	assembler        *ConfigurationAssembler
+	initialNamespace string
+	heartbeatHandler backendapi.HeartbeatHandler
+}
+
+func NewBackend(repository RepositoryFacade, assembler *ConfigurationAssembler,
+	logger *zap.SugaredLogger, initialNamespace string, recorder record.EventRecorder) backendapi.Backend {
+	return &backend{repository: repository,
+		assembler:        assembler,
+		logger:           logger,
+		initialNamespace: initialNamespace,
+		heartbeatHandler: NewSynchronousHandler(repository, recorder, logger)}
+}
+
+func (b *backend) ShouldEdgeDeviceBeUnregistered(ctx context.Context, name, namespace string) (bool, error) {
+	edgeDevice, err := b.repository.GetEdgeDevice(ctx, name, namespace)
+	if err != nil {
+		return false, err
+	}
+
+	if edgeDevice.DeletionTimestamp == nil || utils.HasFinalizer(&edgeDevice.ObjectMeta, YggdrasilWorkloadFinalizer) {
+		return false, nil
+	}
+
+	if utils.HasFinalizer(&edgeDevice.ObjectMeta, YggdrasilConnectionFinalizer) {
+		err = b.repository.RemoveEdgeDeviceFinalizer(ctx, edgeDevice, YggdrasilConnectionFinalizer)
```

Comment:
By the name of the function, I don't expect it to make changes to the manifest. Perhaps the removal of the finalizers should be extracted from here and moved right after the call to this function?

Reply:
Finalizer concept is related to Kubernetes mechanics and as such must not be exposed to the client of `Backend` interface.
I was afraid of what you observed - on one hand the function answers whether device should (must!) trigger deregistration procedure and on the other it has side effects in Kubernetes land.
What do you think about changing the name to something like `FinalizeDeviceRemovalIfNeeded`, `MaybeFinalizeDeviceRemoval` in which case the bool returned from the function can be interpreted as "was the device finally removed"?

Reply:
In that case it is a very specific k8s behavior - caller doesn't and shouldn't know that it needs to make an update of the device. From caller perspective it doesn't matter what happens to the data store, if it's changed or not when `GetXXX` is called.
Maybe if I introduce additional status of `unregistering` it would allow the caller to call something like `FinalizeUnregistration` on the API

Reply:
What bothers me in this case is relying on device accessing the device-server and then on the device-server contacting the backend to invoke a management logic.
My thought is why do we need it here and not have it done by the controller as part of device's reconciliation?

Reply:
In a follow-up PR I will remove the finalizers completely and that should resolve this behavioral inconsistency.